File: haskell/Parse.hs
module Parse (
    parse
) where

import qualified Data.Map as Map
import qualified Data.Maybe as Maybe

import qualified Tokenize
import qualified AST

type Token = Tokenize.Token

parse :: String -> AST.Node
parse raw_string =
    let (node, tokens) = parse_expression (Tokenize.to_tokens raw_string)
    in
        node

parse_expression :: [Token] -> (AST.Node, [Token])
parse_expression [] = error "Hi Eric and Quinten"
parse_expression ((Tokenize.LParen):tokens) = parse_func_call tokens
parse_expression ((Tokenize.IntLiteral int):tokens) = (AST.IntNode int, tokens)
parse_expression ((Tokenize.BoolLiteral bool):tokens) = (AST.BoolNode bool, tokens)
parse_expression ((Tokenize.String_ name):tokens) = (AST.VariableNode name, tokens)

parse_func_call :: [Token] -> (AST.Node, [Token])
parse_func_call ((Tokenize.String_ current):tokens) =
    let maybe_function = Map.lookup current function_map
    in
        if Maybe.isJust maybe_function then
            (Maybe.fromJust maybe_function) tokens
        else
            let (operands, pre_close_tokens) = parse_operands [] tokens
                post_close_tokens = chomp_close_expression pre_close_tokens
            in
                (AST.FunctionCallNode current operands, post_close_tokens)
parse_func_call (token:tokens) =
    error $ show token

parse_if :: [Token] -> (AST.Node, [Token])
parse_if tokens =
    let (cond_expr, tokens1) = parse_expression tokens
        (then_expr, tokens2) = parse_expression tokens1
        (else_expr, tokens3) = parse_expression tokens2
        tokens4 = chomp_close_expression tokens3
    in
        (AST.IfNode cond_expr then_expr else_expr, tokens4)

parse_define :: [Token] -> (AST.Node, [Token])
parse_define ((Tokenize.String_ name): tokens) =
    let (expression, tokens1) = parse_expression tokens
        tokens2 = chomp_close_expression tokens1
        (body, tokens3) = parse_expression tokens2
    in
        (AST.BindingNode name expression body, tokens3)
parse_define (token:tokens) =
    error $ show token

parse_lambda :: [Token] -> (AST.Node, [Token])
parse_lambda tokens =
    let tokens1 = chomp_open_lambda_params tokens
        (params, tokens2) = parse_params [] tokens1
        tokens3 = chomp_close_lambda_params tokens2
        (body, tokens4) = parse_expression tokens3
        tokens5 = chomp_close_expression tokens4
    in
        (AST.LambdaNode body params, tokens5)


parse_params :: [String] -> [Token] -> ([String], [Token])
parse_params existing_params input_tokens@((Tokenize.RBracket):_) =
    (existing_params, input_tokens)
parse_params existing_params ((Tokenize.String_ name):tokens) =
    parse_params (name: existing_params) tokens


parse_operands :: [AST.Node] -> [Token] -> ([AST.Node], [Token])
parse_operands existing_operands input_tokens@((Tokenize.RParen):_) =
    (existing_operands, input_tokens)
parse_operands existing_operands input_tokens =
    let (current, remaining_tokens) = parse_expression input_tokens
    in
        parse_operands (current: existing_operands) remaining_tokens


chomp_close_expression :: [Token] -> [Token]
chomp_close_expression tokens =
    assert_chomping Tokenize.RParen tokens

chomp_open_lambda_params :: [Token] -> [Token]
chomp_open_lambda_params tokens =
    assert_chomping Tokenize.LBracket tokens

chomp_close_lambda_params :: [Token] -> [Token]
chomp_close_lambda_params tokens =
    assert_chomping Tokenize.RBracket tokens

assert_chomping :: Token -> [Token] -> [Token]
assert_chomping expected (token:tokens) =
    if token == expected then
        tokens
    else
        error $ "wrong token'" ++ (show expected) ++ "', found '" ++ (show token) ++ "'"

function_map :: Map.Map String ([Token] -> (AST.Node, [Token]))
function_map = Map.fromList [
    ("if", parse_if),
    ("define", parse_define),
    ("lambda", parse_lambda)]

File: python/src/parse.py
import tokenize
import ast

bools = ['true', 'false']


class Parser:
    def __init__(self, source_string):
        self.tokenizer = tokenize.Tokenizer(source_string)
        self.function_map = {
            'if': self.if_,
            'define': self.define,
            'lambda': self.lambda_,
            'struct': self.struct,
            'member': self.member,
            'with': self.with_,
        }

    def parse_expression(self):
        current = self.tokenizer.chomp()
        if current == '(':
            return self.func_call()
        elif current == tokenize.EOF:
            raise ParsingError("Expecting an expression expecting %s" % tokenize.EOF)
        elif current.isdigit() or (current[1:].isdigit() and current[0] == '-'):
            return ast.IntNode(current)
        elif current in bools:
            return ast.BoolNode(current)
        else:
            return ast.VariableNode(current)

    def maybe_parse_expression(self):
        if self.tokenizer.peek() == tokenize.EOF:
            return ast.GetEnvironmentBindingNode()
        else:
            return self.parse_expression()

    def func_call(self):
        func_name = self.tokenizer.chomp()
        if func_name in self.function_map:
            func = self.function_map[func_name]()
            if func_name not in ('define', 'with'):
                assert_close_expression(self.tokenizer.chomp())
            return func
        else:
            operands = []
            while True:
                if self.tokenizer.peek() == ')':
                    break
                operands.append(self.parse_expression())
            assert_close_expression(self.tokenizer.chomp())
            return ast.FunctionCallNode(func_name, operands)

    def define(self):
        name = self.tokenizer.chomp()
        expression = self.parse_expression()
        assert_close_expression(self.tokenizer.chomp())
        body = self.maybe_parse_expression()
        return ast.BindingNode(name, expression, body)

    def if_(self):
        cond_expr = self.parse_expression()
        then_expr = self.parse_expression()
        else_expr = self.parse_expression()
        return ast.ConditionalNode(cond_expr, then_expr, else_expr)

    def lambda_(self):
        remaining_args = None
        open_bracket = self.tokenizer.chomp()
        args = []
        while True:
            if self.tokenizer.peek() == ']':
                break
            if self.tokenizer.peek() == '...':
                assert_elipsis(self.tokenizer.chomp())
                remaining_args = self.tokenizer.chomp()
                break
            args.append(self.tokenizer.chomp())
        assert_close_params(self.tokenizer.chomp())
        body = self.parse_expression()
        return ast.LambdaNode(body, args, remaining_args)

    def struct(self):
        members = []
        while True:
            if self.tokenizer.peek() == ')':
                break
            members.append(self.tokenizer.chomp())
        return ast.StructDeclarationNode(members)

    def member(self):
        struct_name = self.tokenizer.chomp()
        member_name = self.tokenizer.chomp()
        return ast.MemberAccessNode(struct_name, member_name)

    def with_(self):
        module_name = self.tokenizer.chomp()
        assert_close_expression(self.tokenizer.chomp())
        body = self.maybe_parse_expression()
        return ast.LoadingNode(module_name, body)


class ParsingError(Exception):
    pass


def assert_close_expression(token):
    _assert_is_token(token, ')')


def assert_close_params(token):
    _assert_is_token(token, ']')


def assert_elipsis(token):
    _assert_is_token(token, '...')


def _assert_is_token(to_check, expected):
    if to_check != expected:
        raise ParsingError("Expecting %s found %s" % (expected, to_check))
